File: tech-dhtml/connect4/models/Turn.js
import { HumanPlayer, RandomPlayer } from './Player.js'
import { Color } from './Board.js';

class Turn {

    static #NUMBER_PLAYERS = 2;
    #players;
    #activePlayer;
    #board;


    constructor(board) {
        this.#board = board;
        this.#players = [];
    }
    reset(numberOfHumanPlayers) {
        for (let i = 0; i < Turn.#NUMBER_PLAYERS; i++) {
            this.#players[i] = i < numberOfHumanPlayers ?
                new HumanPlayer(Color.get(i), this.#board) :
                new RandomPlayer(Color.get(i), this.#board);
        }
        this.#activePlayer = 0;
    }
    play(column) {
        this.#players[this.#activePlayer].play(column);
        if (!this.#board.isFinished()) {
            this.#activePlayer = (this.#activePlayer + 1) % Turn.#NUMBER_PLAYERS;
        }
    }
    getActivePlayer() {
        return this.#players[this.#activePlayer];
    }
    getBoard() {
        return this.#board;
    }
    getNumberPlayers() {
        return Turn.#NUMBER_PLAYERS;
    }
    #getNumberOfHumanPlayers(){
        this.numberOfHumanPlayers=0;
        this.#players.forEach((player)=>{player.accept(this)})
        let numberOfHumanPlayers=this.numberOfHumanPlayers
        delete this.numberOfHumanPlayers;
        return numberOfHumanPlayers;
    }
    toJSON(){
        return {
            numberOfHumanPlayers: this.#getNumberOfHumanPlayers(),
            activePlayer: this.#activePlayer
        }
    }
    fromJSON(turn){
        this.reset(turn.numberOfHumanPlayers);
        this.#activePlayer=turn.activePlayer;
    }

    visitHumanPlayer() {
        this.numberOfHumanPlayers++;
    }
    visitRandomPlayer() {
    }
}

export { Turn };

File: tech-java-orientadoObjetos/connecta$/src/connect4/models/Turn.java
package connect4.models;

public class Turn {
    final static int NUMBER_PLAYERS = 2;
    private Player[] players;
    private int activePlayer;
    private Board board;

    public Turn(Board board) {
        this.board = board;
        this.players = new Player[NUMBER_PLAYERS];
    }

    public void reset(Player[] players) {

        this.players = players;
        this.activePlayer = 0;
    }

    public void play(int column) {
        this.players[this.activePlayer].play(column);
        if (!this.board.isGameFinished()) {
            this.activePlayer = (this.activePlayer + 1) % Turn.NUMBER_PLAYERS;
        }
    }

    public Player getActivePlayer() {
        return this.players[this.activePlayer];
    }

    public Board getBoard() {
        return this.board;
    }

    public int getNumberPlayers() {
        return Turn.NUMBER_PLAYERS;
    }
}
